_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+```properties
+# ...
+replica.fetch.max.bytes=1048576
+# ...
+```
+
+The importance of Kafka's topic replication mechanism cannot be overstated.
+Topic replication is central to Kafka's reliability and data durability. Using replication, a failed broker can recover from in-sync replicas on other brokers.
+We go into more detail about leaders, followers and in-sync replicas with [partition rebalancing for availability](#partition-rebalancing-for-availability).
+
+The `auto.create.topics.enable` property is enabled by default to create topics automatically, but it's usually disabled as Kafka users tend to prefer applying more control over topic creation.
+If you do use automatic topic creation, set `num.partitions` to equal the number of brokers in the cluster so that writes are distributed.
+The `delete.topic.enable` property is enabled by default to allow topics to be deleted and is usually disabled too. This time you're making sure that data is not accidentally lost, although you can temporarily enable the property to delete topics if circumstances demand it.
+
+```properties
+# ...
+auto.create.topics.enable=false
+delete.topic.enable=true
+# ...
+```
+
+> You cannot delete topics with the `KafkaTopic` resource if the `delete.topic.enable` property is set to false.
+
+### Internal topic settings for transactions and commits
+
+Pay attention to the configuration requirements for internal Kafka topics.
+
+If you are [using transactions](https://strimzi.io/docs/operators/latest/using.html#reliability_guarantees) to guarantee the reliability of writes to partitions from producers, the internal `__transaction_state` topic used in the process requires a minimum of three brokers with the default settings.
+
+```properties
+# ...
+transaction.state.log.replication.factor=3
+transaction.state.log.min.isr=2
+# ...
+```
+
+The `__consumer_offsets` topic, which stores partition offset commits, has default settings for the number of partitions and replication factor.
+
+```properties
+# ...
+offsets.topic.num.partitions=50
+offsets.topic.replication.factor=3
+offsets.commit.required.acks=-1
+# ...
+```
+
+***:warning: Do not reduce these settings in production.***
+
+### Improving request handling throughput by increasing I/O threads
+
+Network threads (`num.network.threads`) handle requests to the Kafka cluster, such as produce and fetch requests from client applications.
+Adjust the number of network threads to reflect the replication factor and the levels of activity from client producers and consumers interacting with the Kafka cluster.
+
+I/O threads (`num.io.threads`) pick up requests from the request queue to process them. Adding more threads can improve throughput, but the number of CPU cores/and disk bandwidth imposes a practical upper limit.
+The minimum number of threads should equal the number of storage volumes.
```

Comment:
I wonder if we're phrasing this right. It's true that that is a reasonable minimum, but in practice people don't want to scale this down to a minimum, they want to tune it to what's best for their disks, and will happily turn it up. The default is 8, so maybe suggesting that `8*#disks` is a reasonable place to start tuning is more useful that mentioning a minimum.

Reply:
Makes sense. I've changed to:

> A good starting point might be to start with the default of 8 multiplied by the number of disks.